src/loan.rs

```diff
@@ -74,38 +104,107 @@ impl CollateralContract {
         borrower_pk: PublicKey,
         lender_pk: PublicKey,
         timelock: u64,
-        repayment_output: TxOut,
+        (repayment_principal_output, repayment_principal_output_blinder): (TxOut, SecretKey),
+        oracle_pk: PublicKey,
+        min_price_btc: u64,
+        min_timestamp: u64,
     ) -> Result<Self> {
-        let repayment_output = serialize(&repayment_output).to_hex();
+        use elements::opcodes::all::*;
+
+        let repayment_principal_output_as_hex = serialize(&repayment_principal_output).to_hex();
+
+        // the first element of the covenant descriptor is the shared
+        // `covenant_pk`, which is only used to verify that the transaction
+        // data on the witness stack matches the transaction which triggered
+        // the call.
         let descriptor = Descriptor::<elements::bitcoin::PublicKey>::from_str(
             &(format!(
                 "elcovwsh({},or_i(and_v(v:pk({}),after({})),and_v(v:pk({}),outputs_pref({}))))",
-                COVENANT_PK, lender_pk, timelock, borrower_pk, repayment_output,
+                COVENANT_PK, lender_pk, timelock, borrower_pk, repayment_principal_output_as_hex,
             )),
         )
         .context("invalid collateral output descriptor")?;
 
+        let dynamic_liquidation_branch = Builder::new()
+            // check that the lender authorises the spend
+            .push_slice(lender_pk.serialize().as_slice())
+            .push_opcode(OP_CHECKSIGVERIFY)
+            // copy message BTC price to top of stack
+            .push_int(2)
+            .push_opcode(OP_PICK)
+            // copy message timestamp to top of stack
+            .push_int(2)
+            .push_opcode(OP_PICK)
+            // reconstruct message as `timestamp:btc_price`
+            .push_opcode(OP_CAT)
+            // check if oracle has approved of the hashed message (OP_CSFS
+            // implicitly hashes the message)
+            .push_slice(oracle_pk.serialize().as_slice())
+            .push_opcode(OP_CHECKSIGFROMSTACKVERIFY)
+            // check if message is recent enough
+            .push_int(min_timestamp as i64)
+            .push_opcode(OP_GREATERTHANOREQUAL)
+            .push_opcode(OP_VERIFY)
+            // check if BTC price has dipped below minimum
+            .push_int(min_price_btc as i64)
+            .push_opcode(OP_LESSTHANOREQUAL)
+            .into_script();
```

Comment:
Beautiful script :) 

I would like to see this liquidation path evolve like this: 

Ideally it does not matter who _calls_ liquidate on the contract but when liquidating we need to ensure that the lender gets compensated. For that, the liquidator has to payback the debt to the lender but receives in return the bitcoin. 

I've created a discussion for this here: https://github.com/comit-network/baru/discussions/29